source/distributor/service_impl.cc

```diff
@@ -0,0 +1,117 @@
+#include "distributor/service_impl.h"
+
+#include <grpc++/grpc++.h>
+
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "external/envoy/source/common/common/assert.h"
+#include "external/envoy/source/common/protobuf/message_validator_impl.h"
+#include "external/envoy/source/common/protobuf/utility.h"
+
+#include "api/distributor/distributor.pb.validate.h"
+
+namespace Nighthawk {
+
+::grpc::Status NighthawkDistributorServiceImpl::validateRequest(
+    const ::nighthawk::DistributedRequest& request) const {
+  auto& validation_visitor = Envoy::ProtobufMessage::getStrictValidationVisitor();
+  try {
+    Envoy::MessageUtil::validate(request, validation_visitor);
+  } catch (const Envoy::ProtoValidationException& e) {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, e.what());
+  }
+
+  if (request.has_execution_request()) {
+    const ::nighthawk::client::ExecutionRequest& execution_request = request.execution_request();
+    if (!execution_request.start_request().has_options()) {
+      return grpc::Status(
+          grpc::StatusCode::INVALID_ARGUMENT,
+          "DistributedRequest.ExecutionRequest.StartRequest MUST have CommandLineOptions.");
+    }
+  } else {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                        "DistributedRequest.ExecutionRequest MUST be specified.");
+  }
+  return ::grpc::Status::OK;
+}
+
+absl::StatusOr<nighthawk::client::ExecutionResponse>
+NighthawkDistributorServiceImpl::handleExecutionRequest(
+    const envoy::config::core::v3::Address& service,
+    const ::nighthawk::client::ExecutionRequest& request) const {
+  RELEASE_ASSERT(service_client_ != nullptr, "service_client_ != nullptr");
+  std::unique_ptr<nighthawk::client::NighthawkService::Stub> stub;
```

Comment:
can we instead declare this where it's used?

Reply:
done in 6471b8687332bcae77a637c8b1d86640299e20c6